nexus/src/sagas.rs

```diff
@@ -184,72 +184,188 @@ async fn sic_alloc_server(
+            ),
+        },
+        vpc_name: default_name.clone(),
+        subnet_name: default_name.clone(),
+        ip: None, // Request an IP address allocation
+    };
     let vpc = osagactx
         .datastore()
-        .vpc_fetch_by_name(&params.project_id, &default_name)
+        .vpc_fetch_by_name(
+            &saga_params.project_id,
+            &internal_default_name.clone(),
+        )
         .await
         .map_err(ActionError::action_failed)?;
     let subnet = osagactx
         .datastore()
-        .vpc_subnet_fetch_by_name(&vpc.id(), &default_name)
+        .vpc_subnet_fetch_by_name(&vpc.id(), &internal_default_name)
         .await
         .map_err(ActionError::action_failed)?;
 
     let mac = db::model::MacAddr::new().map_err(ActionError::action_failed)?;
     let interface_id = Uuid::new_v4();
-    // Request an allocation
-    let ip = None;
     let interface = db::model::IncompleteNetworkInterface::new(
         interface_id,
         instance_id,
-        // TODO-correctness: vpc_id here is used for name uniqueness. Should
-        // interface names be unique to the subnet's VPC or to the
-        // VPC associated with the instance's default interface?
         vpc.id(),
-        subnet,
+        subnet.clone(),
         mac,
-        ip,
-        params::NetworkInterfaceCreate {
-            identity: IdentityMetadataCreateParams {
-                // By naming the interface after the instance id, we should
-                // avoid name conflicts on creation.
-                name: format!("default-{}", instance_id).parse().unwrap(),
-                description: format!(
-                    "default interface for {}",
-                    params.create_params.identity.name
-                ),
-            },
-        },
-    );
-
+        interface_params.identity.clone(),
+        interface_params.ip,
+    )
+    .map_err(ActionError::action_failed)?;
     let interface = osagactx
         .datastore()
         .instance_create_network_interface(interface)
         .await
+        .map_err(db::subnet_allocation::NetworkInterfaceError::into_external)
         .map_err(ActionError::action_failed)?;
-    Ok(interface.into())
+    Ok(Some(vec![interface.into()]))
 }
 
-async fn sic_create_network_interface_undo(
+async fn sic_create_network_interfaces_undo(
     sagactx: ActionContext<SagaInstanceCreate>,
 ) -> Result<(), anyhow::Error> {
     let osagactx = sagactx.user_data();
-    let network_interface =
-        sagactx.lookup::<NetworkInterface>("network_interface")?;
-
-    osagactx
-        .datastore()
-        .instance_delete_network_interface(&network_interface.identity.id)
-        .await?;
+    let network_interfaces = sagactx
+        .lookup::<Option<Vec<NetworkInterface>>>("network_interfaces")?;
+
+    // No networking at all created
+    if network_interfaces.is_none() {
+        return Ok(());
+    }
+
+    // Collect the parameters of the saga. If we were asked to create
```

Comment:
Ah, sorry that's a stray comment from the previous iteration. There, I was expecting to have NICs laying around, unattached to any instance. The comment was to indicate that we shouldn't delete those.

That no longer applies, since NICs are by construction attached to an instance. I'll remove the comment, thanks for catching it.